producer-c/producer-cloudwatch-integ/CanaryStreamUtils.cpp

```diff
@@ -0,0 +1,228 @@
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    Aws::CloudWatch::Model::Dimension dimension;
+    PCanaryStreamCallbacks pCanaryStreamCallbacks = NULL;
+
+    CHK(ppStreamCallbacks != NULL, STATUS_NULL_ARG);
+
+    // Allocate the entire structure
+    pCanaryStreamCallbacks = (PCanaryStreamCallbacks) MEMCALLOC(1, SIZEOF(CanaryStreamCallbacks));
+    CHK(pCanaryStreamCallbacks != NULL, STATUS_NOT_ENOUGH_MEMORY);
+
+    // Set the version, self
+    pCanaryStreamCallbacks->streamCallbacks.version = STREAM_CALLBACKS_CURRENT_VERSION;
+    pCanaryStreamCallbacks->streamCallbacks.customData = (UINT64) pCanaryStreamCallbacks;
+    pCanaryStreamCallbacks->timeOfNextKeyFrame = new std::map<UINT64, UINT64>();
+
+    pCanaryStreamCallbacks->pCwClient = cwClient;
+
+    dimension.SetName(pStreamName);
+    dimension.SetValue("ProducerSDK");
+
+    pCanaryStreamCallbacks->receivedAckDatum.SetMetricName("ReceivedAckLatency");
+    pCanaryStreamCallbacks->persistedAckDatum.SetMetricName("PersistedAckLatency");
+    pCanaryStreamCallbacks->bufferingAckDatum.SetMetricName("BufferedAckLatency");
+    pCanaryStreamCallbacks->streamErrorDatum.SetMetricName("StreamError");
+    pCanaryStreamCallbacks->currentFrameRateDatum.SetMetricName("FrameRate");
+    pCanaryStreamCallbacks->contentStoreAvailableSizeDatum.SetMetricName("StorageSizeAvailable");
+    pCanaryStreamCallbacks->currentViewDurationDatum.SetMetricName("CurrentViewDuration");
+    pCanaryStreamCallbacks->memoryAllocationSizeDatum.SetMetricName("MemoryAllocation");
+
+    pCanaryStreamCallbacks->receivedAckDatum.AddDimensions(dimension);
+    pCanaryStreamCallbacks->persistedAckDatum.AddDimensions(dimension);
+    pCanaryStreamCallbacks->streamErrorDatum.AddDimensions(dimension);
+    pCanaryStreamCallbacks->currentFrameRateDatum.AddDimensions(dimension);
+    pCanaryStreamCallbacks->contentStoreAvailableSizeDatum.AddDimensions(dimension);
+    pCanaryStreamCallbacks->currentViewDurationDatum.AddDimensions(dimension);
+    pCanaryStreamCallbacks->memoryAllocationSizeDatum.AddDimensions(dimension);
+
+    // Set callbacks
+    pCanaryStreamCallbacks->streamCallbacks.fragmentAckReceivedFn = canaryStreamFragmentAckHandler;
+    pCanaryStreamCallbacks->streamCallbacks.streamErrorReportFn = canaryStreamErrorReportHandler;
+    pCanaryStreamCallbacks->streamCallbacks.freeStreamCallbacksFn = canaryStreamFreeHandler;
+
+CleanUp:
+
+    if (STATUS_FAILED(retStatus)) {
+        pCanaryStreamCallbacks = NULL;
+        MEMFREE(pCanaryStreamCallbacks);
+    }
+
+    // Set the return value if it's not NULL
+    if (ppStreamCallbacks != NULL) {
+        *ppStreamCallbacks = (PStreamCallbacks) pCanaryStreamCallbacks;
+    }
+
+    if(ppCanaryStreamCallbacks != NULL) {
+        *ppCanaryStreamCallbacks = (PCanaryStreamCallbacks) pCanaryStreamCallbacks;
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS freeCanaryStreamCallbacks(PStreamCallbacks* ppStreamCallbacks)
```

Comment:
This can take PCanaryStreamCallbacks

Reply:
That is right. But, keeping it consistent with the way freeContinuousRetryStreamCallbacks is defined in producer C.